### PR TITLE
Rust: added nvptx variant

### PR DESCRIPTION
**Pull request by westernmagic**
@AndrewGaspar

### CODE REVIEW

**Comment by AndrewGaspar** [comment]
Good addition. I have some perhaps mildly annoying feedback, though. I don't think we should do one-off variants for each target architecture you might want to add support for. Rather than an nvptx addition, could we generalize this? Perhaps, add a multi-variant that takes a list of targets to enable?

Something like:
```python
variant('targets', default='', values=('nvptx64-nvidia-cuda'), multi=True, description='Rust target triples to enable')
```

People could add additional target triples as they need them. Some day in the future when Spack can cross-compile to disparate architecture, I presume this will come in handy.

**Comment by westernmagic** [comment]
Implemented support for multiple targets.

I have mixed feelings about this—in general, I agree. On the other hand, the other packages on spack often solve this via single flags for different targets, eg. `gcc`. Then there is also `CudaPackage` which we could inherit from here (as in `llvm`), but I don't think it makes much sense.

**Comment by AndrewGaspar** [comment]
Hm... the `nsimd` package has a multi variant that seems to indicate that you don't need to have a `none` variant. You can just specify nothing:
```python
    variant('optionals',
            default=(),
            description='Optional SIMD features',
            values=('FMA', 'FP16'),
            multi=True)
```

Also, maybe we just don't proscribe any `values` at all? @adamjstewart does this allow any value to be passed?

**Comment by alalazo** [comment]
In case it might be useful, pointing you to a few [helpers](https://spack.readthedocs.io/en/latest/packaging_guide.html#complex-validation-logic-for-variant-values) I made a while ago for complex variants. See for instance `fftw` as an example of use.

**Comment by westernmagic** [comment]
I didn't know about that `nsimd` trick, thanks! Is that supported though? The documentation doesn't talk about it: https://spack.readthedocs.io/en/latest/packaging_guide.html#multi-valued-variants

File: var/spack/repos/builtin/packages/rust/package.py [hunk 1]
```diff
@@ -64,6 +71,7 @@ class Rust(Package):
     depends_on('openssl')
     depends_on('libssh2')
     depends_on('libgit2')
+    depends_on('cuda', when='extra_targets=nvptx-nvidia-cuda')
```
Comment (AndrewGaspar) [inline var/spack/repos/builtin/packages/rust/package.py]: Typo:
```suggestion
    depends_on('cuda', when='extra_targets=nvptx64-nvidia-cuda')
```

File: var/spack/repos/builtin/packages/rust/package.py [hunk 2]
```diff
@@ -55,6 +55,13 @@ class Rust(Package):
         default=True,
         description='Install Rust source files'
     )
+    variant(
+        'extra_targets',
+        default=(),
+        values=('nvptx64-nvidia-cuda',),
```
Comment (AndrewGaspar) [inline var/spack/repos/builtin/packages/rust/package.py]: So I tested, and not specifying `values` seems to work to allow specifying anything, so I think we should maybe just not proscribe allowed values? 

```suggestion
```

File: var/spack/repos/builtin/packages/rust/package.py [hunk 3]
```diff
@@ -55,6 +55,13 @@ class Rust(Package):
...
+        multi=True,
+        description='Triples for extra targets to enable'
```
Comment (AndrewGaspar) [inline var/spack/repos/builtin/packages/rust/package.py]:
```suggestion
        description='Triples for extra targets to enable. For supported targets, see: https://doc.rust-lang.org/nightly/rustc/platform-support.html'
```

File: var/spack/repos/builtin/packages/rust/package.py [hunk 4]
```diff
@@ -496,6 +504,15 @@ def configure(self, spec, prefix):
 
         ar = which('ar', required=True)
 
+        extra_targets = list(self.spec.variants['extra_targets'].value)
+        extra_targets = list(filter(lambda x: x != 'none', extra_targets))
```
Comment (AndrewGaspar) [inline var/spack/repos/builtin/packages/rust/package.py]:
```suggestion
```

File: var/spack/repos/builtin/packages/rust/package.py (hunk 1 above)
Comment (AndrewGaspar) [reply var/spack/repos/builtin/packages/rust/package.py]: Just to check, what does the Rust toolchain need from `cuda`?

**Review by AndrewGaspar — CHANGES_REQUESTED** [review]

File: var/spack/repos/builtin/packages/rust/package.py (hunk 1 above)
Comment (westernmagic) [reply var/spack/repos/builtin/packages/rust/package.py]: AFAIK, calling a global function (`extern "ptx-kernel"`) will not be possible without `cuda`. I guess its not a hard dependency, but I would expect the library to be there when I am targeting PTX

Comment (AndrewGaspar) [reply var/spack/repos/builtin/packages/rust/package.py]: Hm... I guess I disagree with the notion that a package should create a transitive dependency on another package just for convenience. At most, it should be a `type='run'` dependency, but I don't know if I even want to go that far. A consequence of making it a dependency is that switching the `cuda` version you are using would result in you having to rebuild the `rust` package, even though it doesn't functionally change anything about how the `rust` package is built or runs. If the cuda dependency is instead applied to the package that is building and loading the kernels, then you only have to rebuild that one package when building for different versions of cuda.

@alalazo Could you weigh in?

File: var/spack/repos/builtin/packages/rust/package.py [hunk 5]
```diff
@@ -496,6 +503,14 @@ def configure(self, spec, prefix):
 
         ar = which('ar', required=True)
 
+        extra_targets = list(self.spec.variants['extra_targets'].value)
+        targets = [self.get_rust_target()] + extra_targets
+        target_spec = 'target=[' + \
+            ','.join('"{}"'.format(target) for target in targets) + ']'
```
Comment (AndrewGaspar) [inline var/spack/repos/builtin/packages/rust/package.py]: Fix for CI failure

```suggestion
            ','.join('"{0}"'.format(target) for target in targets) + ']'
```

File: var/spack/repos/builtin/packages/rust/package.py [hunk 6]
```diff
@@ -496,6 +503,14 @@ def configure(self, spec, prefix):
...
+        target_specs = '\n'.join(
+            '[target.{}]\nar = "{}"\n'.format(target, ar.path)
```
Comment (AndrewGaspar) [inline var/spack/repos/builtin/packages/rust/package.py]: Fix for CI failure

```suggestion
            '[target.{0}]\nar = "{1}"\n'.format(target, ar.path)
```

File: var/spack/repos/builtin/packages/rust/package.py (hunk 1 above)
Comment (westernmagic) [reply var/spack/repos/builtin/packages/rust/package.py]: That makes sense to me; sorry, its been a long day...
I think the original reason was that `llvm` depends on `cuda` when `+cuda`, but that seems to be only for reasons of offloading using OpenMP, and is therefore not applicable to `rust`

Comment (AndrewGaspar) [reply var/spack/repos/builtin/packages/rust/package.py]: That's alright! Thank you.

**Comment by AndrewGaspar** [comment]
Out of personal curiosity, what kind of projects are you using Rust + CUDA for?

**Comment by westernmagic** [comment]
I am just wrapping my internship up at CSCS: https://www.cscs.ch/about/internships/

Was working on seeing if Rust is useable as a language for HPC :)

**Comment by AndrewGaspar** [comment]
We're getting there... Some resources if you haven't found them already:
- https://github.com/rsmpi/rsmpi
- https://github.com/bheisler/RustaCUDA
- https://lib.rs/crates/ndarray

**Review by AndrewGaspar — APPROVED** [review]

**Comment by AndrewGaspar** [comment]
@adamjstewart @alalazo This looks good from my standpoint.

**Review by adamjstewart — APPROVED** [review]

**Pull request merged**